### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by StupidCodeThief**

### CODE REVIEW

File: client/src/actions/auth.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+  try {
+    const res = await axios.get("http://localhost:5000/api/auth/user");
+
+    console.log(res.data)
+
+    dispatch({
+      type: USER_LOADED,
+      payload: res.data,
+    });
```
Comment (slavasgod) [inline client/src/actions/auth.ts]: Separate all these things on `actionCreators`

File: client/src/actions/auth.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  },
+};
+
+export const loadUser = () => async (dispatch: any): Promise<void> => {
+  if (localStorage.token) {
+    setAuthToken(localStorage.token);
+  }
+
+  try {
+    const res = await axios.get("http://localhost:5000/api/auth/user");
```
Comment (slavasgod) [inline client/src/actions/auth.ts]: Create base url. Actually you can try to create axios http service for requests.

File: client/src/reducers/auth.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  user: null,
+};
+
+export default function (state = initialState, action: any): any {
+  const { payload, type } = action;
+  switch (type) {
+    case REGISTER_SUCCESS:
+    case LOGIN_SUCCESS:
+      localStorage.setItem("token", payload.token);
+      return {
```
Comment (slavasgod) [inline client/src/reducers/auth.ts]: Remove any side effects from reducers, they should be a clear functions

File: client/src/actions/types.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+export const LOGIN_SUCCESS: string = "LOGIN_SUCCESS";
+export const LOGIN_FAILURE: string = "LOGIN_FAILURE";
+
+export const REGISTER_SUCCESS: string = "REGISTER_SUCCESS";
+export const REGISTER_FAILURE: string = "REGISTER_FAILURE";
+
+export const USER_LOADED: string = "USER_LOADED";
+export const AUTH_ERROR: string = "AUTH_ERROR";
+
+export const LOG_OUT: string = "LOG_OUT";
```
Comment (slavasgod) [inline client/src/actions/types.ts]: Create an object, that will contain all of these types

File: client/src/reducers/auth.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  LOG_OUT,
+  AUTH_ERROR,
+  USER_LOADED,
+} from "../actions/types";
+
+interface InitialState {
+  token: string | any;
+  isAuthenticated: boolean;
+  loading: boolean;
+  user: null | any;
```
Comment (slavasgod) [inline client/src/reducers/auth.ts]: Try to avoid `any` and create interface for user

File: client/src/store.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { createStore, applyMiddleware } from "redux";
+import { composeWithDevTools } from "redux-devtools-extension";
+import thunk from "redux-thunk";
+import rootReducer from "./reducers";
+
+const initialState = {};
```
Comment (slavasgod) [inline client/src/store.ts:6]: Create interface for you root state and describe it

File: client/src/reducers/auth.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        loading: false,
+        user: null,
+      };
+    case USER_LOADED:
+      return {
+        ...state,
+        token: localStorage.getItem("token"),
+        isAuthenticated: true,
+        loading: false,
+        user: payload,
```
Comment (slavasgod) [inline client/src/reducers/auth.ts]: Describe all state changes in action payload

**Review by slavasgod — CHANGES_REQUESTED** [review]

File: client/src/components/authentification/Register.tsx [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  },
+}));
+
+const Register: React.FC = () => {
+  const dispatch = useDispatch();
+  const classes = useStyles();
+
+  const isAuthenticated = useSelector(
+    (state: RootStateOrAny) => state.auth.isAuthenticated
+  );
```
Comment (slavasgod) [inline client/src/components/authentification/Register.tsx]: You can create `selectors` folder, where you can write all selectors functions

File: client/src/actions/actions.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
+import axios from "axios";
+
+import actionCreaters from "./actionCreaters";
```
Comment (slavasgod) [inline client/src/actions/actions.ts]: actionCreatOrs

File: client/src/actions/actionCreaters.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+interface IUser {
+  id: number;
+  username: string;
+  email: string;
+}
+
+interface CreaterReturn {
+  type: string;
+  data?: string | IUser;
+}
```
Comment (slavasgod) [inline client/src/actions/actionCreaters.ts]: Migrate to `interfaces.ts` file

**Review by slavasgod — CHANGES_REQUESTED** [review]

File: client/src/store.ts (hunk 6 above)
Comment (StupidCodeThief) [reply client/src/store.ts:6]: done

File: client/src/actions/auth.ts (hunk 1 above)
Comment (StupidCodeThief) [reply client/src/actions/auth.ts]: done

File: client/src/actions/auth.ts (hunk 2 above)
Comment (StupidCodeThief) [reply client/src/actions/auth.ts]: done

File: client/src/reducers/auth.ts (hunk 3 above)
Comment (StupidCodeThief) [reply client/src/reducers/auth.ts]: done

File: client/src/actions/types.ts (hunk 4 above)
Comment (StupidCodeThief) [reply client/src/actions/types.ts]: done

File: client/src/reducers/auth.ts (hunk 5 above)
Comment (StupidCodeThief) [reply client/src/reducers/auth.ts]: done

File: client/src/store.ts (hunk 6 above)
Comment (StupidCodeThief) [reply client/src/store.ts:6]: done

File: client/src/reducers/auth.ts (hunk 7 above)
Comment (StupidCodeThief) [reply client/src/reducers/auth.ts]: done

File: client/src/actions/actionCreaters.ts (hunk 10 above)
Comment (StupidCodeThief) [reply client/src/actions/actionCreaters.ts]: done

File: client/src/actions/actions.ts (hunk 9 above)
Comment (StupidCodeThief) [reply client/src/actions/actions.ts]: done

File: client/src/components/authentification/Register.tsx (hunk 8 above)
Comment (StupidCodeThief) [reply client/src/components/authentification/Register.tsx]: done

**Review by slavasgod — APPROVED** [review]

**Pull request merged**